lib/compress/zstd_compress.c

```diff
@@ -3989,6 +3992,8 @@ size_t ZSTD_compressStream2( ZSTD_CCtx* cctx,
         FORWARD_IF_ERROR( ZSTD_initLocalDict(cctx) , ""); /* Init the local dict if present. */
         ZSTD_memset(&cctx->prefixDict, 0, sizeof(cctx->prefixDict));   /* single usage */
         assert(prefixDict.dict==NULL || cctx->cdict==NULL);    /* only one can be set */
+        if (cctx->cdict)
+            cctx->requestedParams.compressionLevel = cctx->cdict->compressionLevel; /* let cdict take priority in terms of compression level */
```

Comment:
This operation overwrites a parameter set by the user.
We generally try to avoid that, 
it can have surprising impacts in later usages of `cctx` since these parameters are sticky.

That's also the reason for the existence of `appliedParams`,
which is where dynamic adaptations are performed,
there they do not pollute user's explicit requests, nor survive the current session.

Reply:
I've looked into this issue some more - it seems like this behavior also isn't consistent between `ZSTD_compress_usingCDict()` and `ZSTD_compressStream2()`. `ZSTD_compress_usingCDict()` doesn't seem to have this issue, and always will use the CDict compression level.

The reason is: in `ZSTD_compress_usingCDict()`, we essentially have:
```
ZSTD_CCtx_params params = cctx->requestedParams
[...]
params.cParams = ZSTD_getCParamsFromCDict() OR params.cParams = ZSTD_getCParams()
// where both these functions take cdict compression level into account
```
Whereas, in `ZSTD_compressStream2()`, we have
```
ZSTD_CCtx_params params = cctx->requestedParams
params.cParams = ZSTD_getCParamsFromCCtxParams(&cctx->requestedParams, cctx->pledgedSrcSizePlusOne-1, 0 /*dictSize*/));
// followed immediately (within ZSTD_resetCStream_internal) by
params.cParams = ZSTD_getCParamsFromCCtxParams(&params, pledgedSrcSize, dictSize);
// in neither case is cdict compression level taken into account
```

I guess I'm a little perplexed here - what's the point of calling `ZSTD_getCParamsFromCCtxParams()` twice back-to-back, with what are essentially the same arguments (with the exception of dictSize)? Doesn't the second call just overwrite?

Reply:
It's likely redundant in this case.

But `params` is also used for other purposes when `ZSTD_MULTITHREAD` is defined : 
- It makes it possible to adjust `params.nbWorkers` without modifying `cctx->requestedParams`
- It's a parameter of `ZSTDMT_initCStream_internal()`

Also, consider that calling `ZSTD_getCParamsFromCCtxParams()` twice is a very minor impact on performance, making this issue less critical to be adressed.

Nevertheless, if you believe you can improve the situation, you are welcomed.